### ISSUE TITLE
`SkyCoord` gives incorrect results when transforming points near the Declination pole

### ISSUE THREAD

**Issue by bfc5288** [astropy/astropy#16178]
### Description

The `SkyCoord` class provides incorrect Right Ascension values of a coordinate that is located near, or at, a Declination pole (`Dec=90 deg`) when using its internal functions to transform from ICRS to galactic and back to ICRS. No error is given; the value is just incorrect.

### Expected behavior

After creating a `SkyCoord` object in ICRS coordinates that has a Declination at/near a pole and then converting this object to galactic coordinates and back to ICRS, the "new" ICRS coordinates should be equal to the original ICRS coordinates.

### How to Reproduce

Minimal example: start with a point offset from the north Declination pole by `dec_delta`, slowly decrease dec_delta while converting to galactic coordinates and back, and see the resulting Right Ascension value deviate more from the correct value (300 degrees).
```python
>>> from astropy.coordinates import SkyCoord
>>> ra_test = 300; dec_test = 90

>>> dec_delta = 1
>>> icrs_test = SkyCoord(ra_test, dec_test - dec_delta, unit='deg', frame='icrs'); icrs_test
<SkyCoord (ICRS): (ra, dec) in deg
    (300., 89.)>
>>> icrs_test.galactic.icrs    # convert to galactic coordinates and back to ICRS
<SkyCoord (ICRS): (ra, dec) in deg
    (300., 89.)>
# correct

>>> dec_delta = 1E-10
>>> icrs_test = SkyCoord(ra_test, dec_test - dec_delta, unit='deg', frame='icrs'); icrs_test
<SkyCoord (ICRS): (ra, dec) in deg
    (300., 90.)>
>>> icrs_test.galactic.icrs
<SkyCoord (ICRS): (ra, dec) in deg
    (299.99725044, 90.)>
# slightly wrong

>>> dec_delta = 1E-15
>>> icrs_test = SkyCoord(ra_test, dec_test - dec_delta, unit='deg', frame='icrs'); icrs_test
<SkyCoord (ICRS): (ra, dec) in deg
    (300., 90.)>
>>> icrs_test.galactic.icrs
<SkyCoord (ICRS): (ra, dec) in deg
    (291.41843232, 90.)>
# very wrong

>>> dec_delta = 0
>>> icrs_test = SkyCoord(ra_test, dec_test - dec_delta, unit='deg', frame='icrs'); icrs_test
<SkyCoord (ICRS): (ra, dec) in deg
    (300., 90.)>
>>> icrs_test.galactic.icrs
<SkyCoord (ICRS): (ra, dec) in deg
    (291.41843232, 90.)>
# problem remains at the pole
```

Note that this doesn't seem to be a problem when near/at the Right Ascension pole:
```python
>>> ra_test = 360
>>> ra_delta = 1E-15
>>> icrs_test = SkyCoord(ra_test - ra_delta, dec_test, unit='deg', frame='icrs'); icrs_test
<SkyCoord (ICRS): (ra, dec) in deg
    (0., 80.)>
>>> icrs_test.galactic.icrs
<SkyCoord (ICRS): (ra, dec) in deg
    (360., 80.)>
# correct since 0=360 for periodic RA

>>> ra_delta = 0
>>> icrs_test = SkyCoord(ra_test, dec_test, unit='deg', frame='icrs'); icrs_test
<SkyCoord (ICRS): (ra, dec) in deg
    (0., 80.)>
>>> icrs_test.galactic.icrs
<SkyCoord (ICRS): (ra, dec) in deg
    (360., 80.)>
```

The issue persists when using the underlying function that `SkyCoord.galactic` (or `SkyCoord.icrs`) use (`SkyCoord.transform_to(...)`). It is present for astropy versions 4, 5, and 6. I haven't tested other coordinate systems yet since I don't use them in my code.

### Versions

```
Linux-6.5.0-21-generic-x86_64-with-glibc2.35
Python 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:50:58) [GCC 12.3.0]
astropy 6.0.0
Numpy 1.26.4
pyerfa 2.0.1.1
Scipy not installed
Matplotlib not installed
```

**Comment by eerovaher** [comment]
It is an inherent feature of spherical coordinate systems that their two poles have coordinate singularities, which cause the longitude values at the poles to be meaningless. This is not limited to the ICRS, nor is it caused by the `astropy` implementation of it. If you look at the separation of the original and round-tripped coordinates you will see that it is very small:
```python
>>> from astropy.coordinates import SkyCoord
>>> ra_test = 300
>>> dec_test = 90
>>> dec_delta = 1E-15
>>> icrs_test = SkyCoord(ra_test, dec_test - dec_delta, unit="deg")
>>> icrs_test
<SkyCoord (ICRS): (ra, dec) in deg
    (300., 90.)>
>>> icrs_test.galactic.icrs
<SkyCoord (ICRS): (ra, dec) in deg
    (296.26324045, 90.)>
>>> icrs_test.separation(icrs_test.galactic.icrs)
<Angle 1.27317308e-14 deg>
```
Such a small difference can be explained by numerical inaccuracies that should be expected when working with floating point numbers and it does not look like it is revealing any bugs.

> Note that this doesn't seem to be a problem when near/at the Right Ascension pole:

There is no such thing as Right Ascension pole, just like the Earth has no East pole.

**Comment by dhomeier** [comment]
> Such a small difference can be explained by numerical inaccuracies that should be expected when working with floating point numbers and it does not look like it is revealing any bugs.

You can actually see that already in your input value, as the relative deviation is _O(1e-15/1e2)_:
```python
>>> dec_in = dec_test - dec_delta
>>> dec_in
90.0
>>> np.finfo(dec_in)
finfo(resolution=1e-15, min=-1.7976931348623157e+308, max=1.7976931348623157e+308, dtype=float64)
```

**Comment by bfc5288** [comment]
Thank you for the responses. I agree that this kind of behavior shouldn't be abnormal at the poles since they are general coordinate singularities.

The underlying issue I noticed _might_ be occurring more generally near the poles rather than only at the poles; I just used the poles for the minimal example since it was simpler to write up. I see now that that was a mistake, so I will work on getting a more general minimal example to demonstrate the problem (or will discover that the problem disappears along the way and will close this issue :slightly_smiling_face: ).
